godtools/App/Features/ToolDetails/Presentation/ToolDetails/ToolDetailsViewModel.swift

```diff
@@ -0,0 +1,252 @@
+//
+//  ToolDetailsViewModel.swift
+//  godtools
+//
+//  Created by Levi Eggert on 6/10/22.
+//  Copyright © 2022 Cru. All rights reserved.
+//
+
+import UIKit
+import SwiftUI
+
+class ToolDetailsViewModel: NSObject, ObservableObject {
+    
+    private let resource: ResourceModel
+    private let dataDownloader: InitialDataDownloader
+    private let languageSettingsService: LanguageSettingsService
+    private let localizationServices: LocalizationServices
+    private let favoritedResourcesCache: FavoritedResourcesCache
+    private let analytics: AnalyticsContainer
+    private let getToolTranslationsUseCase: GetToolTranslationsUseCase
+    private let languagesRepository: LanguagesRepository
+    
+    private weak var flowDelegate: FlowDelegate?
+    
+    @Published var mediaType: ToolDetailsMediaType
+    @Published var name: String = ""
+    @Published var totalViews: String = ""
+    @Published var openToolButtonTitle: String = ""
+    @Published var learnToShareToolButtonTitle: String = ""
+    @Published var addToFavoritesButtonTitle: String = ""
+    @Published var removeFromFavoritesButtonTitle: String = ""
+    @Published var hidesLearnToShareToolButton: Bool = true
+    @Published var hidesAddToFavoritesButton: Bool = false
+    @Published var hidesRemoveFromFavoritesButton: Bool = true
+    @Published var aboutTitle: String = ""
+    @Published var versionsTitle: String = ""
+    @Published var aboutDetails: String = ""
+    
+    init(flowDelegate: FlowDelegate, resource: ResourceModel, dataDownloader: InitialDataDownloader, languageSettingsService: LanguageSettingsService, localizationServices: LocalizationServices, favoritedResourcesCache: FavoritedResourcesCache, analytics: AnalyticsContainer, getToolTranslationsUseCase: GetToolTranslationsUseCase, languagesRepository: LanguagesRepository) {
+        
+        self.flowDelegate = flowDelegate
+        self.resource = resource
+        self.dataDownloader = dataDownloader
+        self.languageSettingsService = languageSettingsService
+        self.localizationServices = localizationServices
+        self.favoritedResourcesCache = favoritedResourcesCache
+        self.analytics = analytics
+        self.getToolTranslationsUseCase = getToolTranslationsUseCase
+        self.languagesRepository = languagesRepository
+        self.mediaType = ToolDetailsViewModel.getMediaType(resource: resource, dataDownloader: dataDownloader)
+        
+        super.init()
+        
+        reloadToolDetails()
+        reloadFavorited()
+        setupBinding()
+        reloadLearnToShareToolButtonState()
+    }
+    
+    deinit {
+        print("x deinit: \(type(of: self))")
+        favoritedResourcesCache.resourceFavorited.removeObserver(self)
+        favoritedResourcesCache.resourceUnfavorited.removeObserver(self)
+    }
+    
+    private var analyticsScreenName: String {
```

Comment:
Hey @rachaelblue I think that would be a great idea. 
Maybe too we have a protocol that all of our concrete ViewModel's must implement, something like ViewModelType, (this would be noted in programming guide in the ReadMe), and then anything we want to standardize across all viewModels it's just a matter of adopting that to our ViewModelType protocol.  So if we want all ViewModels to implement PageViewed protocol, just drop that on the ViewModelType. Thoughts?

Reply:
We could run into issues with ViewModelType including analytics, since some of the subview viewModels don't do their own analytics (or are just part of a page and wouldn't have a pageViewed event).  But I do like the idea of having a ViewModelType, I'm just not sure if there's anything that stands out as being true for _all_ viewModels.  Worth thinking about though!

Reply:
That's a good point about the sub viewModels.  Let's scratch the ViewModelType and not make things too complicated.

Reply:
With use case extensions we can create a SharedUseCaseType+PageViewed extension which performs the analytics tracking call.
```func pageViewed(pageViewedAnalytics: PageViewedAnalytics, properties: TrackScreenModel)```.

Reply:
I'm going to think about this some more, I know I would like to have a use case for this logic that way it isn't duplicated across viewModels.